src/Microsoft.ML/Runtime/EntryPoints/TrainTestMacro.cs

```diff
@@ -23,8 +24,14 @@ public sealed class SubGraphInput
 
         public sealed class SubGraphOutput
         {
-            [Argument(ArgumentType.Required, HelpText = "The model", SortOrder = 1)]
-            public Var<IPredictorModel> Model;
+            [Argument(ArgumentType.AtMostOnce, HelpText = "The predictor model", SortOrder = 1)]
+            public Var<IPredictorModel> PredictorModel;
+            
+            [Argument(ArgumentType.AtMostOnce, HelpText = "Transform model", SortOrder = 2)]
+            public Var<ITransformModel> TransformModel;
+
+            [Argument(ArgumentType.AtMostOnce, HelpText = "Indicates to use transform model instead of predictor model.", SortOrder = 3)]
```

Comment:
Why is this necessary? There is a convention for the graph variables: if set to null/absent don't use it. 
Is there any case where the TransformModel is provided but it shouldn't get used?

If so, please add a comment to code.

Reply:
The code generated entrypoint class always initializes the variables(see csharpapi.cs in this PR), so a user may set TransformModel but PredictorModel would also been initialized to a non null value. Hence I cannot know which one the user had set. As of today the user has to explicitly set variables to null that they don't want the code to use for example in TrainTestMacro arguments TransformModel needs to be set to null explicitly each time a graph is created even though in the code it is initialized to null. 

You could argue why not adopt the same model of making the user set values to null that they don't want the code to use and that might be a fair argument as it would clean up the code a little.


---
In reply to: [191048298](https://github.com/dotnet/machinelearning/pull/212#discussion_r191048298) [](ancestors = 191048298)